diff --git a/lib/dns/include/dns/keyvalues.h b/lib/dns/include/dns/keyvalues.h
index 2737e23a1c..ec2226d1d6 100644
--- a/lib/dns/include/dns/keyvalues.h
+++ b/lib/dns/include/dns/keyvalues.h
@@ -68,11 +68,11 @@
 #define DNS_KEYALG_ECDSA384	     14
 #define DNS_KEYALG_ED25519	     15
 #define DNS_KEYALG_ED448	     16
-#define DNS_KEYALG_FALCON512	     17
 #define DNS_KEYALG_DILITHIUM2	     18
 #define DNS_KEYALG_SPHINCSSHA256128S 19
 #define DNS_KEYALG_XMSS		     20
 #define DNS_KEYALG_XMSSMT	     21
+#define DNS_KEYALG_FALCON512        251
 #define DNS_KEYALG_INDIRECT	     252
 #define DNS_KEYALG_PRIVATEDNS	     253
 #define DNS_KEYALG_PRIVATEOID	     254 /*%< Key begins with OID giving alg */
diff --git a/lib/dns/include/dst/dst.h b/lib/dns/include/dst/dst.h
index 4c4feceedc..4593a1cbda 100644
--- a/lib/dns/include/dst/dst.h
+++ b/lib/dns/include/dst/dst.h
@@ -95,7 +95,6 @@ typedef enum dst_algorithm {
 	DST_ALG_ECDSA384 = 14,
 	DST_ALG_ED25519 = 15,
 	DST_ALG_ED448 = 16,
-	DST_ALG_FALCON512 = 17,
 	DST_ALG_DILITHIUM2 = 18,
 	DST_ALG_SPHINCSSHA256128S = 19,
 	DST_ALG_XMSS = 20,
@@ -115,7 +114,7 @@ typedef enum dst_algorithm {
 	DST_ALG_HMACSHA384 = 164, /* XXXMPA */
 	DST_ALG_HMACSHA512 = 165, /* XXXMPA */
 	DST_ALG_HMAC_LAST = DST_ALG_HMACSHA512,
-
+	DST_ALG_FALCON512 = 251,
 	DST_ALG_INDIRECT = 252,
 	DST_ALG_PRIVATE = 254,
 	DST_MAX_ALGS = 256,
