diff --git a/lib/dns/include/dns/keyvalues.h b/lib/dns/include/dns/keyvalues.h
index 3a520aea8b..6a275ca6f2 100644
--- a/lib/dns/include/dns/keyvalues.h
+++ b/lib/dns/include/dns/keyvalues.h
@@ -112,7 +112,8 @@
 #define DNS_SIG_MAYO2SIZE 180
 #define DNS_KEY_MAYO2SIZE 5488
 
-#define DNS_SIG_FALCON512SIZE 666
+/* #define DNS_SIG_FALCON512SIZE 666  # padded*/
+#define DNS_SIG_FALCON512SIZE 752
 #define DNS_KEY_FALCON512SIZE 897
 
 #define DNS_SIG_DILITHIUM2SIZE 2420
diff --git a/lib/dns/openssloqs_link.c b/lib/dns/openssloqs_link.c
index f135f4b2d3..5c8e58c3ab 100644
--- a/lib/dns/openssloqs_link.c
+++ b/lib/dns/openssloqs_link.c
@@ -58,7 +58,7 @@ static const oqs_alginfo_t *
 openssloqs_alg_info(dst_algorithm_t key_alg) {
 	if (key_alg == DST_ALG_FALCON512) {
 		static const oqs_alginfo_t oqs_alginfo = {
-			.alg_name = "falconpadded512",
+			.alg_name = "falcon512",
 			.key_size = DNS_KEY_FALCON512SIZE,
 			.priv_key_size = FALCON512_PRIVATEKEYSIZE,
 			.sig_size = DNS_SIG_FALCON512SIZE,
