diff --git a/bin/dnssec/dnssec-keyfromlabel.c b/bin/dnssec/dnssec-keyfromlabel.c
index 80022095b1..167d8bc5d6 100644
--- a/bin/dnssec/dnssec-keyfromlabel.c
+++ b/bin/dnssec/dnssec-keyfromlabel.c
@@ -398,6 +398,7 @@ main(int argc, char **argv) {
 			case DST_ALG_ECDSA384:
 			case DST_ALG_ED25519:
 			case DST_ALG_ED448:
+			case DST_ALG_MAYO2:
 			case DST_ALG_FALCON512:
 			case DST_ALG_DILITHIUM2:
 			case DST_ALG_SPHINCSSHA256128S:
diff --git a/bin/dnssec/dnssec-keygen.c b/bin/dnssec/dnssec-keygen.c
index d352739984..160ebd9004 100644
--- a/bin/dnssec/dnssec-keygen.c
+++ b/bin/dnssec/dnssec-keygen.c
@@ -163,7 +163,7 @@ usage(void) {
 	fprintf(stderr, "        ECDSAP256SHA256 | ECDSAP384SHA384 |\n");
 	fprintf(stderr, "        ED25519 | ED448\n");
 	fprintf(stderr, "        ED25519 | ED448 | DH\n");
-	fprintf(stderr, "        FALCON512 | DILITHIUM2 |\n");
+	fprintf(stderr, "        FALCON512 | DILITHIUM2 | MAYO2 \n");
 	fprintf(stderr, "        SPHINCS+-SHA256-128S\n");
 	fprintf(stderr, "    -3: use NSEC3-capable algorithm\n");
 	fprintf(stderr, "    -b <key size in bits>:\n");
@@ -181,6 +181,7 @@ usage(void) {
 	fprintf(stderr, "        ED448:\tignored\n");
 	fprintf(stderr, "        FALCON512:\tignored\n");
 	fprintf(stderr, "        DILITHIUM2:\tignored\n");
+	fprintf(stderr, "        MAYO2:\tignored\n");
 	fprintf(stderr, "        SPHINCS+-SHA256-128S:\tignored\n");
 	fprintf(stderr, "        (key size defaults are set according to\n"
 			"        algorithm and usage (ZSK or KSK)\n");
@@ -367,6 +368,7 @@ keygen(keygen_ctx_t *ctx, isc_mem_t *mctx, int argc, char **argv) {
 			case DST_ALG_ECDSA384:
 			case DST_ALG_ED25519:
 			case DST_ALG_ED448:
+			case DST_ALG_MAYO2:
 			case DST_ALG_FALCON512:
 			case DST_ALG_DILITHIUM2:
 			case DST_ALG_SPHINCSSHA256128S:
@@ -422,6 +424,7 @@ keygen(keygen_ctx_t *ctx, isc_mem_t *mctx, int argc, char **argv) {
 			case DST_ALG_ECDSA384:
 			case DST_ALG_ED25519:
 			case DST_ALG_ED448:
+			case DST_ALG_MAYO2:
 			case DST_ALG_FALCON512:
 			case DST_ALG_DILITHIUM2:
 			case DST_ALG_SPHINCSSHA256128S:
@@ -591,6 +594,9 @@ keygen(keygen_ctx_t *ctx, isc_mem_t *mctx, int argc, char **argv) {
 	case DST_ALG_ED448:
 		ctx->size = 456;
 		break;
+	case DST_ALG_MAYO2:
+		ctx->size = 5488*8;
+		break;
 	case DST_ALG_FALCON512:
 		ctx->size = 7176;
 		break;
diff --git a/lib/dns/dst_api.c b/lib/dns/dst_api.c
index 9a301f199c..19f63ef84f 100644
--- a/lib/dns/dst_api.c
+++ b/lib/dns/dst_api.c
@@ -230,6 +230,7 @@ dst_lib_init(isc_mem_t *mctx, const char *engine) {
 #if HAVE_GSSAPI
 	RETERR(dst__gssapi_init(&dst_t_func[DST_ALG_GSSAPI]));
 #endif /* HAVE_GSSAPI */
+	RETERR(dst__openssloqs_init(&dst_t_func[DST_ALG_MAYO2]));
 	RETERR(dst__openssloqs_init(&dst_t_func[DST_ALG_FALCON512]));
 	RETERR(dst__openssloqs_init(&dst_t_func[DST_ALG_DILITHIUM2]));
 	RETERR(dst__openssloqs_init(&dst_t_func[DST_ALG_SPHINCSSHA256128S]));
@@ -1490,6 +1491,9 @@ dst_key_sigsize(const dst_key_t *key, unsigned int *n) {
 	case DST_ALG_GSSAPI:
 		*n = 128; /*%< XXX */
 		break;
+	case DST_ALG_MAYO2:
+		*n = DNS_SIG_MAYO2SIZE;
+		break;
 	case DST_ALG_FALCON512:
 		*n = DNS_SIG_FALCON512SIZE;
 		break;
diff --git a/lib/dns/dst_parse.c b/lib/dns/dst_parse.c
index 6ae0f45e5e..d87ef4d622 100644
--- a/lib/dns/dst_parse.c
+++ b/lib/dns/dst_parse.c
@@ -109,6 +109,9 @@ static struct parse_map map[] = {
 	{ TAG_HMACSHA512_KEY, "Key:" },
 	{ TAG_HMACSHA512_BITS, "Bits:" },
 
+	{ TAG_MAYO2_PRIVATEKEY, "PrivateKey:" },
+	{ TAG_MAYO2_PUBLICKEY, "PublicKey:" },
+
 	{ TAG_FALCON512_PRIVATEKEY, "PrivateKey:" },
 	{ TAG_FALCON512_PUBLICKEY, "PublicKey:" },
 
@@ -468,6 +471,7 @@ check_data(const dst_private_t *priv, const unsigned int alg, bool old,
 		return (check_hmac_sha(priv, HMACSHA384_NTAGS, alg));
 	case DST_ALG_HMACSHA512:
 		return (check_hmac_sha(priv, HMACSHA512_NTAGS, alg));
+	case DST_ALG_MAYO2:
 	case DST_ALG_FALCON512:
 	case DST_ALG_DILITHIUM2:
 	case DST_ALG_SPHINCSSHA256128S:
@@ -817,6 +821,9 @@ dst__privstruct_writefile(const dst_key_t *key, const dst_private_t *priv,
 	case DST_ALG_HMACSHA512:
 		fprintf(fp, "(HMAC_SHA512)\n");
 		break;
+	case DST_ALG_MAYO2:
+		fprintf(fp, "(MAYO2)\n");
+		break;
 	case DST_ALG_FALCON512:
 		fprintf(fp, "(FALCON512)\n");
 		break;
diff --git a/lib/dns/dst_parse.h b/lib/dns/dst_parse.h
index 2fae6a4edd..529a71b56b 100644
--- a/lib/dns/dst_parse.h
+++ b/lib/dns/dst_parse.h
@@ -96,6 +96,8 @@
 #define TAG_HMACSHA512_BITS ((DST_ALG_HMACSHA512 << TAG_SHIFT) + 1)
 
 #define OQS_PQC_NTAGS		  2
+#define TAG_MAYO2_PRIVATEKEY  ((DST_ALG_MAYO2 << TAG_SHIFT) + 0)
+#define TAG_MAYO2_PUBLICKEY	  ((DST_ALG_MAYO2 << TAG_SHIFT) + 1)
 #define TAG_FALCON512_PRIVATEKEY  ((DST_ALG_FALCON512 << TAG_SHIFT) + 0)
 #define TAG_FALCON512_PUBLICKEY	  ((DST_ALG_FALCON512 << TAG_SHIFT) + 1)
 #define TAG_DILITHIUM2_PRIVATEKEY ((DST_ALG_DILITHIUM2 << TAG_SHIFT) + 0)
diff --git a/lib/dns/include/dns/keyvalues.h b/lib/dns/include/dns/keyvalues.h
index ec2226d1d6..3a520aea8b 100644
--- a/lib/dns/include/dns/keyvalues.h
+++ b/lib/dns/include/dns/keyvalues.h
@@ -72,6 +72,7 @@
 #define DNS_KEYALG_SPHINCSSHA256128S 19
 #define DNS_KEYALG_XMSS		     20
 #define DNS_KEYALG_XMSSMT	     21
+#define DNS_KEYALG_MAYO2            249
 #define DNS_KEYALG_FALCON512        251
 #define DNS_KEYALG_INDIRECT	     252
 #define DNS_KEYALG_PRIVATEDNS	     253
@@ -108,6 +109,9 @@
 #define DNS_KEY_ED25519SIZE 32
 #define DNS_KEY_ED448SIZE   57
 
+#define DNS_SIG_MAYO2SIZE 180
+#define DNS_KEY_MAYO2SIZE 5488
+
 #define DNS_SIG_FALCON512SIZE 666
 #define DNS_KEY_FALCON512SIZE 897
 
diff --git a/lib/dns/include/dst/dst.h b/lib/dns/include/dst/dst.h
index 4593a1cbda..a22f8102e0 100644
--- a/lib/dns/include/dst/dst.h
+++ b/lib/dns/include/dst/dst.h
@@ -114,6 +114,7 @@ typedef enum dst_algorithm {
 	DST_ALG_HMACSHA384 = 164, /* XXXMPA */
 	DST_ALG_HMACSHA512 = 165, /* XXXMPA */
 	DST_ALG_HMAC_LAST = DST_ALG_HMACSHA512,
+	DST_ALG_MAYO2 = 249,
 	DST_ALG_FALCON512 = 251,
 	DST_ALG_INDIRECT = 252,
 	DST_ALG_PRIVATE = 254,
@@ -121,7 +122,8 @@ typedef enum dst_algorithm {
 } dst_algorithm_t;
 
 /*% A buffer of this size is large enough to hold any key */
-#define DST_KEY_MAXSIZE 5120 // OQS changed from 1280
+// #define DST_KEY_MAXSIZE 5120 // OQS changed from 1280
+#define DST_KEY_MAXSIZE 5492 // OQS+MAYO2 changed from 5120
 
 /*%
  * A buffer of this size is large enough to hold the textual representation
diff --git a/lib/dns/kasp.c b/lib/dns/kasp.c
index e083354f68..54ca1e0c74 100644
--- a/lib/dns/kasp.c
+++ b/lib/dns/kasp.c
@@ -456,6 +456,9 @@ dns_kasp_key_size(dns_kasp_key_t *key) {
 	case DNS_KEYALG_ED448:
 		size = 456;
 		break;
+	case DST_ALG_MAYO2:
+		size = 5488*8;
+		break;
 	case DST_ALG_FALCON512:
 		size = 7176;
 		break;
diff --git a/lib/dns/openssloqs_link.c b/lib/dns/openssloqs_link.c
index acf51ff3d7..f135f4b2d3 100644
--- a/lib/dns/openssloqs_link.c
+++ b/lib/dns/openssloqs_link.c
@@ -39,6 +39,7 @@
 		goto err; \
 	}
 
+#define MAYO2_PRIVATEKEYSIZE	 	 24
 #define FALCON512_PRIVATEKEYSIZE	 1281
 #define DILITHIUM2_PRIVATEKEYSIZE	 2528
 #define SPHINCSSHA256128S_PRIVATEKEYSIZE 64
@@ -69,6 +70,19 @@ openssloqs_alg_info(dst_algorithm_t key_alg) {
 		};
 		return &oqs_alginfo;
 	}
+	if (key_alg == DST_ALG_MAYO2) {
+		static const oqs_alginfo_t oqs_alginfo = {
+			.alg_name = "mayo2",
+			.key_size = DNS_KEY_MAYO2SIZE,
+			.priv_key_size = MAYO2_PRIVATEKEYSIZE,
+			.sig_size = DNS_SIG_MAYO2SIZE,
+			.tags = {
+				.ntags = OQS_PQC_NTAGS,
+				.private_key_tag = TAG_MAYO2_PRIVATEKEY,
+				.public_key_tag = TAG_MAYO2_PUBLICKEY, },
+		};
+		return &oqs_alginfo;
+	}
 	if (key_alg == DST_ALG_DILITHIUM2) {
 		static const oqs_alginfo_t oqs_alginfo = {
 			.alg_name = "dilithium2",
@@ -515,11 +529,13 @@ openssloqs_parse(dst_key_t *key, isc_lex_t *lexer, dst_key_t *pub) {
 
 	for (i = 0; i < priv.nelements; i++) {
 		switch (priv.elements[i].tag) {
+		case TAG_MAYO2_PRIVATEKEY:
 		case TAG_FALCON512_PRIVATEKEY:
 		case TAG_DILITHIUM2_PRIVATEKEY:
 		case TAG_SPHINCSSHA256128S_PRIVATEKEY:
 			privkey_index = i;
 			break;
+		case TAG_MAYO2_PUBLICKEY:
 		case TAG_FALCON512_PUBLICKEY:
 		case TAG_DILITHIUM2_PUBLICKEY:
 		case TAG_SPHINCSSHA256128S_PUBLICKEY:
diff --git a/lib/dns/rcode.c b/lib/dns/rcode.c
index ef49791f98..094ecdc3cb 100644
--- a/lib/dns/rcode.c
+++ b/lib/dns/rcode.c
@@ -111,6 +111,7 @@
 		{ DNS_KEYALG_ECDSA384, "ECDSA384", 0 },                      \
 		{ DNS_KEYALG_ED25519, "ED25519", 0 },                        \
 		{ DNS_KEYALG_ED448, "ED448", 0 },                            \
+		{ DNS_KEYALG_MAYO2, "MAYO2", 0 },                            \
 		{ DNS_KEYALG_FALCON512, "FALCON512", 0 },                    \
 		{ DNS_KEYALG_DILITHIUM2, "DILITHIUM2", 0 },                  \
 		{ DNS_KEYALG_SPHINCSSHA256128S, "SPHINCS+-SHA256-128S", 0 }, \
diff --git a/lib/isccfg/kaspconf.c b/lib/isccfg/kaspconf.c
index a18d879f50..f687d1fc18 100644
--- a/lib/isccfg/kaspconf.c
+++ b/lib/isccfg/kaspconf.c
@@ -238,6 +238,7 @@ cfg_kaspkey_fromconfig(const cfg_obj_t *config, dns_kasp_t *kasp,
 			case DNS_KEYALG_ECDSA384:
 			case DNS_KEYALG_ED25519:
 			case DNS_KEYALG_ED448:
+			case DNS_KEYALG_MAYO2:
 			case DNS_KEYALG_FALCON512:
 			case DNS_KEYALG_DILITHIUM2:
 			case DNS_KEYALG_SPHINCSSHA256128S:
