commit 2aeb0420392282d062feeb6831ae885d08ce2b6c
Author: Caspar Schutijser <caspar.schutijser@sidn.nl>
Date:   Wed Jun 4 11:39:21 2025 +0200

    Adapt MAYO2 support to round 2 version of MAYO

diff --git a/bin/dnssec/dnssec-keygen.c b/bin/dnssec/dnssec-keygen.c
index 160ebd9004..61b55222e2 100644
--- a/bin/dnssec/dnssec-keygen.c
+++ b/bin/dnssec/dnssec-keygen.c
@@ -595,7 +595,7 @@ keygen(keygen_ctx_t *ctx, isc_mem_t *mctx, int argc, char **argv) {
 		ctx->size = 456;
 		break;
 	case DST_ALG_MAYO2:
-		ctx->size = 5488*8;
+		ctx->size = 4912*8;
 		break;
 	case DST_ALG_FALCON512:
 		ctx->size = 7176;
diff --git a/lib/dns/include/dns/keyvalues.h b/lib/dns/include/dns/keyvalues.h
index 3a520aea8b..f031312809 100644
--- a/lib/dns/include/dns/keyvalues.h
+++ b/lib/dns/include/dns/keyvalues.h
@@ -109,8 +109,8 @@
 #define DNS_KEY_ED25519SIZE 32
 #define DNS_KEY_ED448SIZE   57
 
-#define DNS_SIG_MAYO2SIZE 180
-#define DNS_KEY_MAYO2SIZE 5488
+#define DNS_SIG_MAYO2SIZE 186
+#define DNS_KEY_MAYO2SIZE 4912
 
 #define DNS_SIG_FALCON512SIZE 666
 #define DNS_KEY_FALCON512SIZE 897
diff --git a/lib/dns/include/dst/dst.h b/lib/dns/include/dst/dst.h
index a22f8102e0..a6ae6e284d 100644
--- a/lib/dns/include/dst/dst.h
+++ b/lib/dns/include/dst/dst.h
@@ -123,7 +123,8 @@ typedef enum dst_algorithm {
 
 /*% A buffer of this size is large enough to hold any key */
 // #define DST_KEY_MAXSIZE 5120 // OQS changed from 1280
-#define DST_KEY_MAXSIZE 5492 // OQS+MAYO2 changed from 5120
+// #define DST_KEY_MAXSIZE 5492 // OQS+MAYO2 changed from 5120
+#define DST_KEY_MAXSIZE 5120 // OQS+MAYO2,round2 change it back
 
 /*%
  * A buffer of this size is large enough to hold the textual representation
diff --git a/lib/dns/kasp.c b/lib/dns/kasp.c
index 54ca1e0c74..294434d7d2 100644
--- a/lib/dns/kasp.c
+++ b/lib/dns/kasp.c
@@ -457,7 +457,7 @@ dns_kasp_key_size(dns_kasp_key_t *key) {
 		size = 456;
 		break;
 	case DST_ALG_MAYO2:
-		size = 5488*8;
+		size = 4912*8;
 		break;
 	case DST_ALG_FALCON512:
 		size = 7176;
